teps/0044-decouple-task-composition-from-scheduling.md

```diff
@@ -0,0 +1,954 @@
+co-locate pods that use the same workspace.
+
+Cons:
+* Doesn't help us with the overhead of multiple pods
+* [TEP-0046](https://github.com/tektoncd/community/pull/318/files) explores shortcomings of this approach
+
+### Custom scheduler
+
+In this approach we use a custom scheduler to schedule our pods.
+
+Cons:
+* [@jlpetterson has explored this option](https://docs.google.com/document/d/1lIqFP1c3apFwCPEqO0Bq9j-XCDH5XRmq8EhYV_BPe9Y/edit#heading=h.18c0pv2k7d1a)
+  and felt it added more complexity without much gain (see also https://github.com/tektoncd/pipeline/issues/3052)
+* Doesn't help us with the overhead of multiple pods (each Task would still be a pod)
+
+### Support other ways to share data (e.g. buckets)
+
+In this approach we could add more workspace types that support other ways of sharing data between pods, for example
+uploading to and downloading from s3.
+
+[This is something we support for "linking" PipelineResources.](https://github.com/tektoncd/pipeline/blob/master/docs/install.md#configuring-pipelineresource-storage)
+
+Cons:
+* Uploading and downloading at the beginning and end of every Task is not as efficient as being able to share the same
+  disk
+* We'd need to define an extension mechanism so folks can use whatever backing store they want
+* Doesn't help us with the overhead of multiple pods (each Task would still be a pod)
+
+### Focus on workspaces
+
+This approach assumes that our main issue is around how to deal with data in a workspace:
+
+1. Before we run our logic
+2. After we're done with our logic
+
+(Where "logic" means whatever we're actually trying to do; i.e. getting the data to act on and doing something with the
+results are not the main concern of whatever our Pipeline is doing.)
+
+In the following example, the first thing the Pipeline will do is run `git-clone` to initialize the contents of the
+`source-code` workspace; after the Pipeline finishes executing, `gcs-upload` will be called to do whatever is needed
+with the data in the `test-results` workspace.
+
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: Pipeline
+metadata:
+  name: build-test-deploy
+spec:
+ params:
+  - name: url
+    value: https://github.com/tektoncd/pipeline.git
+  - name: revision
+    value: v0.11.3
+ workspaces:
+   - name: source-code
+     init:
+     - taskRef: git-clone
+       params:
+       - name: url
+         value: https://github.com/tektoncd/pipeline.git
+       - name: revision
+         value: v0.11.3
+   - name: test-results
+     teardown:
+     - taskRef: gcs-upload
+       params:
+       - name: location
+         value: gs://my-test-results-bucket/testrun-$(taskRun.name)
+ tasks:
+ - name: run-unit-tests
+   taskRef:
+     name: just-unit-tests
+   workspaces:
+    - name: source-code
+    - name: test-results
+```
+
+Pros:
+* Solves one specific problem: getting data on and off of workspaces
+
+Cons:
```

Comment:
I think another *con* might be another kind of overhead (push/pull at each task, … might slow the pipeline quite a bit)

Reply:
oh interesting! i assumed this would be just once for the entire pipeline, but i see what you mean, it could be per task using the workspace - ill add a note about this